### PR TITLE
[MM-20501] Migrated 'components/post_view/post_body' to ts/tsx

### PR DESCRIPTION
**Pull request by vipul08**
#### Summary

This pull request includes migration of js/jsx to ts/tsx for components/post_view/post_body' module
#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/15440

#### Related Pull Requests

- Has redux changes (https://github.com/mattermost/mattermost-redux/commit/8f99a5dcf9a923a97afba54ca248a6c771027912)

### CODE REVIEW

**Review by Willyfrog — CHANGES_REQUESTED** [review]
Hi there! thanks for your contribution! It seems that the linting failed, can you run `make check-style` and fix the complains from the linter?

**Comment by vipul08** [comment]
Sure I will do that 👍

**Comment by devinbinnie** [comment]
@vipul08 Thanks for the contribution!
Looks like you have a type-check error that needs to be resolved. Once this is resolved, I'll have a look.

```
> mattermost-webapp@0.0.1 check-types /root/mattermost/mattermost-webapp
> tsc

components/post_view/post_body/post_body.tsx:196:21 - error TS2322: Type '{ post: Post; compactDisplay: boolean | undefined; hasMention: boolean; }' is not assignable to type 'IntrinsicAttributes & Pick<Pick<ClassAttributes<PostMessageView> & Props, "post" | "key" | "ref" | "options" | "isRHS" | "compactDisplay">, "post" | ... 2 more ... | "compactDisplay"> & Partial<...> & Partial<...>'.
  Property 'hasMention' does not exist on type 'IntrinsicAttributes & Pick<Pick<ClassAttributes<PostMessageView> & Props, "post" | "key" | "ref" | "options" | "isRHS" | "compactDisplay">, "post" | ... 2 more ... | "compactDisplay"> & Partial<...> & Partial<...>'.

196                     hasMention={true}
                        ~~~~~~~~~~


Found 1 error.
```

**Comment by vipul08** [comment]
@Willyfrog @devinbinnie - Fixed the type error.

File: components/post_view/post_body/index.ts [hunk 1]
```diff
@@ -3,13 +3,28 @@
...
-import PostBody from './post_body.jsx';
+import {Post} from 'mattermost-redux/types/posts';
+import {GlobalState} from 'mattermost-redux/types/store';
 
-function mapStateToProps(state, ownProps) {
+import {PluginsState} from 'types/store/plugins';
+
+import PostBody from './post_body';
+
+interface State extends GlobalState {
```
Comment (devinbinnie) [inline components/post_view/post_body/index.ts:20]: We have actually have a version of `GlobalState` declared in the webapp that includes this, so you don't need this interface.
You can import it from `types/store`.

File: components/post_view/post_body/post_body.tsx [hunk 2]
```diff
@@ -17,89 +20,93 @@ import PostBodyAdditionalContent from 'components/post_view/post_body_additional
...
+
+    /**
+   * Flag passed down to PostBodyAdditionalContent for determining if post embed is visible
+   */
+    isEmbedVisible?: boolean;
+
+    /**
+   * Whether or not the post username can be overridden.
+   */
+    enablePostUsernameOverride?: boolean;
```
Comment (devinbinnie) [inline components/post_view/post_body/post_body.tsx:74]: This is required in the original component.

File: components/post_view/post_body/post_body.tsx [hunk 3]
```diff
@@ -17,89 +20,93 @@ import PostBodyAdditionalContent from 'components/post_view/post_body_additional
...
+   * The post to render the body of
+   */
+    post: Post;
+
+    parentPost?: Post;
+
+    /**
+   * The parent post of the thread this post is in
+   */
+    parentPostUser?: UserProfile | null;
```
Comment (devinbinnie) [inline components/post_view/post_body/post_body.tsx:39]: Are we checking for `null` specifically at some point for this property? I couldn't find it myself.

File: components/post_view/post_message_view/post_message_view.tsx [hunk 4]
```diff
@@ -27,6 +27,7 @@ type Props = {
     theme: Theme; /* Logged in user's theme */
     pluginPostTypes?: any; /* Post type components from plugins */
     currentRelativeTeamUrl: string;
+    hasMention?: boolean;
```
Comment (devinbinnie) [inline components/post_view/post_message_view/post_message_view.tsx:30]: `hasMention` doesn't appear to be used by this component, so I think we can remove it here and remove the property passed in `PostBody`

File: package-lock.json [hunk 5]
```diff
@@ -9511,14 +9511,14 @@
       "dependencies": {
         "file-type": {
           "version": "3.9.0",
-          "resolved": "https://registry.npmjs.org/file-type/-/file-type-3.9.0.tgz",
```
Comment (devinbinnie) [inline package-lock.json:9514]: Can we revert these changes to `package-lock.json`? They aren't relevant to this PR.

**Review by devinbinnie — APPROVED** [review]
Thanks for the contribution @vipul08!
I've added a few comments for you to review.

File: components/post_view/post_body/index.ts [hunk 6]
```diff
@@ -18,7 +33,8 @@ function mapStateToProps(state, ownProps) {
     }
 
     const config = getConfig(state);
-    const enablePostUsernameOverride = config.EnablePostUsernameOverride === 'true';
+    const enablePostUsernameOverride =
+    config.EnablePostUsernameOverride === 'true';
```
Comment (Willyfrog) [inline components/post_view/post_body/index.ts:37]: why do we have a newline here?

File: components/post_view/commented_on/commented_on.tsx [hunk 7]
```diff
@@ -17,7 +17,7 @@ import UserProfile from '../../user_profile/user_profile';
 type Props = {
     displayName?: string;
     enablePostUsernameOverride?: boolean;
-    parentPostUser?: UserProfileType;
+    parentPostUser?: UserProfileType | null;
```
Comment (Willyfrog) [inline components/post_view/commented_on/commented_on.tsx:20]: doesn't having the `?` operator already allows values to be `null`?

File: components/post_view/post_body/post_body.tsx [hunk 8]
```diff
@@ -146,7 +158,7 @@ export default class PostBody extends React.PureComponent {
         }
 
         let failedOptions;
-        if (this.props.post.failed) {
+        if (this.props.post && this.props.post.failed) {
```
Comment (Willyfrog) [inline components/post_view/post_body/post_body.tsx:161]: wouldn't this be more idiomatic like 
```suggestion
        if (this.props.post!.failed) {
```

File: components/post_view/post_body/post_body.tsx [hunk 9]
```diff
@@ -212,15 +240,12 @@ export default class PostBody extends React.PureComponent {
             <div>
                 {comment}
                 <div
-                    id={`${post.id}_message`}
+                    id={`${post ? post.id : ''}_message`}
```
Comment (Willyfrog) [inline components/post_view/post_body/post_body.tsx:243]: do we want to have `_message` as an id?

**Review by Willyfrog — CHANGES_REQUESTED** [review]

**Comment by Willyfrog** [comment]
hey @vipul08 let us know if you need any help with the reviews. Also notice that there are some conflicts now that need to be solved.

In that regard, can you also remove package-lock from the list of submitted files? we don't usually accept it if there is no change in `package.json` to prevent many conflicts that would arise with it.

**Comment by vipul08** [comment]
@Willyfrog - Thanks for reaching out!

I am working on the review comments and will resolve it by tomorrow.

File: components/post_view/post_body/index.ts (hunk 1 above)
Comment (vipul08) [reply components/post_view/post_body/index.ts:20]: I have imported from `import {GlobalState} from 'mattermost-redux/types/store';` here only. Are you suggesting something else?

File: components/post_view/post_body/post_body.tsx (hunk 2 above)
Comment (vipul08) [reply components/post_view/post_body/post_body.tsx:74]: Which component are you referring to ? @devinbinnie

File: components/post_view/post_body/post_body.tsx (hunk 3 above)
Comment (vipul08) [reply components/post_view/post_body/post_body.tsx:39]: If I will not put this then compiler start complaining in the `index.ts` file. 

image

Comment (vipul08) [reply components/post_view/post_body/post_body.tsx:39]: This is the error -->

`Types of parameters 'props' and 'nextProps' are incompatible.
          Type 'Readonly<Matching<{ parentPost: Post | undefined; parentPostUser: UserProfile | null | undefined; pluginPostTypes: { [postType: string]: PostPluginComponent; }; enablePostUsernameOverride: boolean; isReadOnly: boolean; } & DispatchProp<...>, ClassAttributes<...> & Props>>' is not assignable to type 'Props'.
            Types of property 'parentPostUser' are incompatible.
              Type 'UserProfile | null | undefined' is not assignable to type 'UserProfile | undefined'.
                Type 'null' is not assignable to type 'UserProfile | undefined'.ts(2345)`

File: components/post_view/commented_on/commented_on.tsx (hunk 7 above)
Comment (vipul08) [reply components/post_view/commented_on/commented_on.tsx:20]: No, it will assume the `parentPostUser` will only have `UserProfileType` and undefined not null

File: components/post_view/post_body/post_body.tsx (hunk 9 above)
Comment (vipul08) [reply components/post_view/post_body/post_body.tsx:243]: @Willyfrog - I am not sure about it. I only added null check into it. Please suggest what should we add here.

**Comment by vipul08** [comment]
@Willyfrog @devinbinnie - I have added few questions on to your comments. Could you please have a look?

File: components/post_view/post_body/post_body.tsx (hunk 9 above)
Comment (Willyfrog) [reply components/post_view/post_body/post_body.tsx:243]: I'd define a constant outside of the div
```js
const divId = post ? `${post.id}_message` : undefined
```
and then the div:
```
<div
    id={divId}
```
that way we don't have the option of repeating IDs

it's either that, or having a different unique identifier that we could rely on when there is no id in the post.

File: components/post_view/post_body/index.ts (hunk 1 above)
Comment (devinbinnie) [reply components/post_view/post_body/index.ts:20]: Yes, you should use the type provided by `types/store`:

`import {GlobalState} from 'types/store'`

File: components/post_view/post_body/post_body.tsx (hunk 2 above)
Comment (devinbinnie) [reply components/post_view/post_body/post_body.tsx:74]: In `post_body.jsx`:

```
        /**
         * Whether or not the post username can be overridden.
         */
        enablePostUsernameOverride: PropTypes.bool.isRequired,
```

File: components/post_view/post_body/post_body.tsx (hunk 3 above)
Comment (devinbinnie) [reply components/post_view/post_body/post_body.tsx:39]: Ahh, it's because of this line in index.js:
`parentPostUser = parentPost ? getUser(state, parentPost.user_id) : null;`

We need to change that `null` to `undefined` to fix that.

File: components/post_view/post_body/index.ts (hunk 1 above)
Comment (vipul08) [reply components/post_view/post_body/index.ts:20]: I was getting this error thats why I declared the type `State` @devinbinnie 


image

File: components/post_view/post_body/post_body.tsx (hunk 2 above)
Comment (vipul08) [reply components/post_view/post_body/post_body.tsx:74]: I guess we can not add Proptypes directly in .ts file. Can I make `enablePostUsernameOverride ?: boolean` to `enablePostUsernameOverride : boolean`?

File: components/post_view/post_body/index.ts (hunk 1 above)
Comment (devinbinnie) [reply components/post_view/post_body/index.ts:20]: Is your branch synced with `master`? We added the `plugins` field quite recently to that type AFAIK.

File: components/post_view/post_body/post_body.tsx (hunk 2 above)
Comment (devinbinnie) [reply components/post_view/post_body/post_body.tsx:74]: Yes that should be correct.

File: components/post_view/post_body/index.ts (hunk 1 above)
Comment (vipul08) [reply components/post_view/post_body/index.ts]: I tried to sync the master... even in the `store.d.ts` there is no `plugins` property in `GlobalState`

**Comment by vipul08** [comment]
@Willyfrog @devinbinnie - Please review the changes

**Review by Willyfrog — CHANGES_REQUESTED** [review]
Can you remove `package-lock.json`? if we are not making changes in package.json it is safer not to add it to the changes.

Added a couple of comments about formating

File: components/post_view/post_body/post_body.tsx [hunk 10]
```diff
@@ -156,7 +168,12 @@ export default class PostBody extends React.PureComponent {
         }
 
         let fileAttachmentHolder = null;
-        if (((post.file_ids && post.file_ids.length > 0) || (post.filenames && post.filenames.length > 0)) && this.props.post.state !== Posts.POST_DELETED) {
+        if (
+            ((post && post.file_ids && post.file_ids.length > 0) ||
+        (post && post.filenames && post.filenames.length > 0)) &&
+      this.props.post &&
+      this.props.post.state !== Posts.POST_DELETED
```
[the remark this answers is not in the document]
Comment (vipul08) [reply components/post_view/post_body/post_body.tsx]: @Willyfrog - Somehow I  am not able to remove the `package-lock.json`. I tried few things but couldn't remove it from the PR. 
I tried these steps:

`git reset HEAD^ package-lock.json `
`git commit --amend --no-edit`
`git push -f`

[the remark this answers is not in the document]
Comment (Willyfrog) [reply components/post_view/post_body/post_body.tsx]: since you didn't add it the previous commit, that won't work.

you can probably check it out from your local master to fix it:
`git checkout master -- package-lock.json`

File: components/post_view/post_body/index.ts (hunk 1 above)
Comment (devinbinnie) [reply components/post_view/post_body/index.ts]: There is right here:
https://github.com/mattermost/mattermost-webapp/blob/eafb2f4b394f1d1788f1038b3ea19f0da32b6805/types/store/index.ts#L17

It links to another file if that helps.

File: components/post_view/post_body/post_body.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (vipul08) [reply components/post_view/post_body/post_body.tsx]: I tried this... but seems like it didn't remove from the PR

**Comment by Willyfrog** [comment]
@vipul08 are you ok with me adding a commit to remove the package-lock?

**Comment by vipul08** [comment]
@Willyfrog - Sure. Please do let me know how you will do it?

**Comment by Willyfrog** [comment]
so, to fix this I updated my master branch first, then checked out the file I needed and merged with master:
```sh
git fetch --all
git checkout upstream/master -- package-lock.json
git merge upstream/master
```

**Comment by vipul08** [comment]
@Willyfrog Thank you. I will take a note of it for future reference 👍

**Comment by devinbinnie** [comment]
Hey @vipul08, thanks for the contribution!

I'm seeing a few TS compiler errors still lingering, let us know if you need any help fixing any of them.

```
components/post_view/post_body/index.ts:38:32 - error TS2339: Property 'plugins' does not exist on type 'GlobalState'.

38         pluginPostTypes: state.plugins.postTypes,
                                  ~~~~~~~

components/post_view/post_body/post_body.tsx:172:15 - error TS2532: Object is possibly 'undefined'.

172             ((post?.file_ids?.length > 0) ||
                  ~~~~~~~~~~~~~~~~~~~~~~

components/post_view/post_body/post_body.tsx:173:10 - error TS2532: Object is possibly 'undefined'.

173         (post?.filenames?.length > 0)) &&
             ~~~~~~~~~~~~~~~~~~~~~~~

components/post_view/post_body_additional_content/post_body_additional_content.test.tsx:351:16 - error TS2532: Object is possibly 'undefined'.

351         expect(baseProps.actions.toggleEmbedVisibility).toHaveBeenCalledTimes(1);
                   ~~~~~~~~~~~~~~~~~

components/post_view/post_body_additional_content/post_body_additional_content.test.tsx:352:16 - error TS2532: Object is possibly 'undefined'.

352         expect(baseProps.actions.toggleEmbedVisibility).toBeCalledWith('post_id_1');
                   ~~~~~~~~~~~~~~~~~

components/post_view/post_body_additional_content/post_body_additional_content.tsx:30:9 - error TS2532: Object is possibly 'undefined'.

30         this.props.actions.toggleEmbedVisibility(this.props.post.id);
```

**Comment by devinbinnie** [comment]
Hey @vipul08, do you still plan to finish this PR up?

**Comment by vipul08** [comment]
@devinbinnie - yes  I am planning to finish it in a week. I am continuously traveling so, couldn't get time to look into it.

**Comment by vipul08** [comment]
@devinbinnie @Willyfrog - I am trying to setup again on my new laptop and found lots of compilation error ( might related to node version) . So, I was thinking to merge the master branch to my forked branch but for that it asking me to open up a PR. Please suggest.

**Comment by devinbinnie** [comment]
> @devinbinnie @Willyfrog - I am trying to setup again on my new laptop and found lots of compilation error ( might related to node version) . So, I was thinking to merge the master branch to my forked branch but for that it asking me to open up a PR. Please suggest.

@vipul08 You should be able to just pull down `master`, merge it into your branch, fix conflicts and push to this PR. I don't think you need a new PR for that. It would probably be worth doing all that on your own PC and pushing up, shouldn't be an issue. If you need help with that let us know.

As for Node version, we are on Node v16 now so you'll want to upgrade to that as well.

**Comment by devinbinnie** [comment]
Closing as stale.
